### PR TITLE
Fix NumericReturnTypeFromStrictScalarReturnsRector for non-natively-typed parameters

### PR DESCRIPTION
**Pull request by staabm**

### CODE REVIEW

File: rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php [hunk 1]
```diff
@@ -141,7 +153,7 @@ private function refactorBinaryOp(
         $leftType = $this->getType($binaryOp->left);
         $rightType = $this->getType($binaryOp->right);
 
-        if ($leftType instanceof IntegerType && $rightType instanceof IntegerType) {
+        if ($leftType->isInteger()->yes() && $rightType->isInteger()->yes()) {
```
Comment (staabm) [inline rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: how should this rector behave according to phpdoc typed params?

does the "strict" in the rector name mean it should only trust native types?

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: I am thinking about a test like

```php
<?php

namespace Rector\Tests\TypeDeclaration\Rector\ClassMethod\NumericReturnTypeFromStrictScalarReturnsRector\Fixture;

final class SkipPhpDoc
{
    /**
     * @param 0|1 $first
     */
    public function resolve($first, int $second)
    {
        return $first + $second;
    }
}

```

Comment (TomasVotruba) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: Docblocks are skipped as non-strict types.

Comment (TomasVotruba) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: Could you add test fixture with this code sample to make sure its still skipped?

Comment (samsonasik) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: There is ExprAnalyzer for detect variable from non typed param for that

https://github.com/rectorphp/rector-src/blob/93a279d3658fea5d56670bb63ff41df3b2b481f3/src/NodeAnalyzer/ExprAnalyzer.php#L36

Comment (samsonasik) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: @staabm could you verify latest change of `ExprAnalyzer->isNonTypedFromParam()` works on this? Thank you.

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: it seems it works - thank you

File: rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php [hunk 2]
```diff
@@ -134,14 +157,27 @@ private function matchRootReturnWithExpr(ClassMethod|Function_ $functionLike): ?
...
+
     private function refactorBinaryOp(
         BinaryOp $binaryOp,
         ClassMethod|Function_ $functionLike
     ): null|Function_|ClassMethod {
         $leftType = $this->getType($binaryOp->left);
         $rightType = $this->getType($binaryOp->right);
 
-        if ($leftType instanceof IntegerType && $rightType instanceof IntegerType) {
+        if ($leftType->isInteger()->yes() && $rightType->isInteger()->yes()) {
```
Comment (TomasVotruba) [inline rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: How can we verify this takes into account only direct integers? I'd prefer to be explicit and keep/extend the original code, instead of hoping it avoid docblock integer types, that are interprested as strict types by PHPStan.

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: I thought I did by adding more tests. Do you have something concrete in mind which should be tested in addition?

I think we can't just add `instanceof IntegerRangeType` because there is also a union of integers in the game.

Comment (TomasVotruba) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: I recall doc typed properties and external method/property calls are false positives. Could you add those too?

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: I reverted my implementation changes for now, as you are right in that phpdoc typed properties are a problem.
[these are a problem even before the changes I suggested before this initial PR.](https://getrector.com/demo/ce464291-430f-452d-9846-ddfd4ad894b8)

I re-calibrated the goal of this PR to just add more test-cases for now, so this knowledge about edge cases of this rector is not lost.

----

I got the feeling the only real way this rector could work reliably is if it would use [treatPhpDocTypesAsCertain: false](https://phpstan.org/config-reference#treatphpdoctypesascertain). is this something we could configure on a per rector basis?

Comment (samsonasik) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: Rector can specify phpstan file to be used:

```php
$rectorConfig->phpstanConfig(__DIR__ . '/phpstan-for-rector.neon');
```

Comment (staabm) [reply rules/TypeDeclaration/Rector/ClassMethod/NumericReturnTypeFromStrictScalarReturnsRector.php]: I guess this setting would then affect all rector rules not just a one.

I wonder whether rector should have a "mode" in which it does not at all trust phpdoc, so it can be very sure about the things it is doing. just applying `treatPhpDocTypesAsCertain: false` would regress a lot of existing functionality I guess.

**Comment by staabm** [comment]
I have reduced the PR to the important test-cases as I think the initial goal would [require some things we can't do](https://github.com/rectorphp/rector-src/pull/4380#discussion_r1250388747)?

**Review by TomasVotruba — APPROVED** [review]

**Comment by TomasVotruba** [comment]
Thanks 👍

Lets merge this as it is and iterate from there. Checking only variables or disabling doc types might be an option.

**Comment by staabm** [comment]
> disabling doc types might be an option

how can this be achieved?

**Comment by samsonasik** [comment]
@staabm I think that need failing test case first, to check the rule that affected, and we can see possible solution to compare type vs native type. On this case, verify if the data is property fetch, then check property type itself via Reflection.

**Comment by TomasVotruba** [comment]
@staabm I was referring to PHPStan option:  https://github.com/rectorphp/rector-src/pull/4380#discussion_r1250405455
Not sure how it's implemented in PHPStan, it's been couple of years since I looked into this parameter.
